sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1106,12 +1106,15 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
     // Push [[Filter]] operators through [[Window]] operators. Parts of the predicate that can be
     // pushed beneath must satisfy the following two conditions:
     // 1. All the expressions are part of window partitioning key. The expressions can be compound.
-    // 2. Deterministic
+    // 2. Deterministic.
+    // 3. Placed before any non-deterministic predicates.
     case filter @ Filter(condition, w: Window)
         if w.partitionSpec.forall(_.isInstanceOf[AttributeReference]) =>
       val partitionAttrs = AttributeSet(w.partitionSpec.flatMap(_.references))
+      var isPredicatePushdownAble = true
       val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition { cond =>
-        cond.references.subsetOf(partitionAttrs) && cond.deterministic &&
+        isPredicatePushdownAble = isPredicatePushdownAble && cond.deterministic
+        isPredicatePushdownAble && cond.references.subsetOf(partitionAttrs) &&
```

Comment:
And we should move the `partitionAttrs.forall(_.isInstanceOf[Attribute])` predicate out of the closure.